trunk/source/IcfgToChc/src/de/uni_freiburg/informatik/ultimate/plugins/icfgtochc/concurrent/ThreadModularHornClauseProvider.java

```diff
@@ -0,0 +1,843 @@
+	 *            transition must have an entry; other thread instances must not.
+	 *
+	 *            Specifying multiple predecessors can be useful for forks.
+	 * @param updatedThread
+	 *            The primary active thread, whose local variables are updated by the transition.
+	 */
+	protected HornClauseBuilder buildInductivityClause(final IcfgEdge transition,
+			final Map<ThreadInstance, IcfgLocation> preds, final Map<ThreadInstance, IcfgLocation> succs,
+			final ThreadInstance updatedThread) {
+		final var clause = createBuilder(mInvariantPredicate,
+				"inductivity clause for transition " + transition.hashCode() + " with transformula "
+						+ transition.getTransformula() + " and thread instance " + updatedThread);
+
+		// add body clause
+		clause.addBodyPredicate(mInvariantPredicate, clause.getDefaultBodyArgs(mInvariantPredicate));
+
+		// add location constraints
+		for (final var instance : mInstances) {
+			final var isActive = preds.containsKey(instance) || succs.containsKey(instance);
+			if (isActive) {
+				// if instance only in preds or only in succs, the respective other location is mBottomLocation
+				clause.differentBodyHeadVar(mLocationVars.get(instance));
+				addInLocationConstraint(clause, instance, preds.get(instance));
+				addOutLocationConstraint(clause, instance, succs.get(instance));
+			}
+		}
+
+		// add transition constraint
+		final var locals = mLocalVars.values().collect(Collectors.toList());
+		addTransitionConstraint(clause, transition, updatedThread, locals);
+
+		return clause;
+	}
+
+	protected List<HornClauseBuilder> buildNonInterferenceClauses(final IcfgEdge transition) {
+		final var result = new ArrayList<HornClauseBuilder>();
+		result.add(buildNonInterferenceClause(transition));
+		return result;
+	}
+
+	/**
+	 * Builds a noninterference clause for the given transition.
+	 */
+	protected HornClauseBuilder buildNonInterferenceClause(final IcfgEdge transition) {
+		// TODO support transitions with multiple predecessors (joins)
+		final var interferingThread = getInterferingThread(transition);
+		final var interferingVars = createThreadSpecificVars(interferingThread);
+
+		return buildNonInterferenceClause(transition, (clause, instance) -> {
+			final var bodyArgs = new ArrayList<>(mInvariantPredicate.getParameters());
+			replaceThreadVariables(bodyArgs, instance, interferingVars);
+			return bodyArgs.stream().map(v -> clause.getBodyVar(v).getTerm()).collect(Collectors.toList());
+		});
+	}
+
+	protected HornClauseBuilder buildNonInterferenceClause(final IIcfgTransition<?> transition,
+			final BiFunction<HornClauseBuilder, ThreadInstance, List<Term>> getInterferencePrecondition) {
+		final var clause = createBuilder(mInvariantPredicate, "non-interference clause for transition "
+				+ transition.hashCode() + " with transformula " + transition.getTransformula());
+
+		// TODO support transitions with multiple predecessors (joins)
```

Comment:
Does this TODO mean that we just ignore joins and thus overapproximate the behavior of the program here?